station-items-v0/src/main/java/net/modificationstation/stationapi/api/bonemeal/BonemealAPI.java

```diff
@@ -0,0 +1,101 @@
+        BlockRegistry.INSTANCE.forEach(blockBase ->
+                blockBase.getStateManager().getStates().stream().filter(state -> state.isIn(ground)).forEach(state -> addPlant(state, plant, weight))
+        );
+    }
+
+    public static void addPlant(TagKey<BlockBase> ground, Structure plant, int weight) {
+        BlockRegistry.INSTANCE.forEach(blockBase ->
+                blockBase.getStateManager().getStates().stream().filter(state -> state.isIn(ground)).forEach(state -> addPlant(state, plant, weight))
+        );
+    }
+
+    public static boolean generate(Level level, int x, int y, int z, BlockState state, int side) {
+        WeightedList<Structure> structures = PLACERS.get(state);
+        if (structures == null) return false;
+        Random random = level.rand;
+        Direction offset = Direction.byId(side);
+        structures.get(random).generate(
+                level,
+                random,
+                x + offset.getOffsetX(),
+                y + offset.getOffsetY(),
+                z + offset.getOffsetZ()
+        );
+        for (byte i = 0; i < 127; i++) {
+            int px = x + random.nextInt(7) - 3;
+            int py = y + random.nextInt(5) - 2;
+            int pz = z + random.nextInt(7) - 3;
+            state = level.getBlockState(px, py, pz);
+            structures = PLACERS.get(state);
+            if (structures == null) continue;
+            structures.get(random).generate(level, random, px, py + 1, pz);
+        }
+        return true;
+    }
+
+    private static class SimpleStateStructure extends Structure {
+        private final BlockState state;
+
+        private SimpleStateStructure(BlockState state) {
+            this.state = state;
+        }
+
+        @Override
+        public boolean generate(Level level, Random random, int x, int y, int z) {
+            BlockState levelState = level.getBlockState(x, y, z);
+            if (!levelState.isAir() && !levelState.getMaterial().isLiquid()) return false;
+            if (state.getBlock().canPlaceAt(level, x, y, z)) {
+                level.setBlockState(x, y, z, state);
+                return true;
+            }
+            return false;
+        }
+    }
+
+    private static class GrassStructure extends Structure {
+        private static final BlockState STATE = BlockBase.TALLGRASS.getDefaultState();
+
+        @Override
+        public boolean generate(Level level, Random random, int x, int y, int z) {
+            BlockState levelState = level.getBlockState(x, y, z);
+            if (!levelState.isAir() && !levelState.getMaterial().isLiquid()) return false;
+            if (STATE.getBlock().canPlaceAt(level, x, y, z)) {
+                level.setBlockState(x, y, z, STATE);
+                level.setTileMeta(x, y, z, 1);
+                return true;
+            }
+            return false;
+        }
+    }
+
+    static {
+        addPlant(BlockBase.GRASS.getDefaultState(), new GrassStructure(), 1);
```

Comment:
Yes, it replaces vanilla generator since calling API cancels vanilla code

Reply:
Flowers no longer generate because of this

Reply:
Now it should be fine

![2023-10-30_16 15 20](https://github.com/ModificationStation/StationAPI/assets/17707380/ffeefdbf-7f83-4a29-8573-984190cd8de7)

Reply:
Don't really understand why only grass generator is replaced out of all bonemeal generators

Reply:
This is the only one generator that spreads blocks. All others are just block modificators and can be overridden with `onBonemealUse` in target blocks. Grass generator was overridden to keep grass density aligned with other plants (since otherwise it will generate grass with wrong density)

Reply:
If you will not override it then:
- If event is not cancellable custom plant will generate with too hight density and override vanilla grass. If its generation chance will be forced to be low it will be not possible to generate hight density plants
- If event is cancellable then adding custom plant will stop any grass from spawning

Basically adding these generators just makes things simple - you get a complete control over custom plant density together with grass, and it will generate patches almost identical to vanilla

Reply:
When you say "it will generate patches almost identical to vanilla", do you mean with custom plants or without?

Reply:
Both, by default it will generate vanilla grass and flowers in vanilla way, but you can add more flowers/grasses (or even structures) to generation pool, and they can have weight even larger than vanilla grass

Reply:
I'm kind of hesitant of overriding vanilla generation like this. Though, to be fair, the original code does use `ItemBase#rand` which uses a different seed each run, so it *probably* won't affect anything.

Reply:
That produces a result equal to vanilla code. It is possible to use more injects into vanilla code, but there will be no difference except worse performance and troubles for mods to mixin above that